### ISSUE TITLE
Make Server Exception give more details than superficial details

### ISSUE THREAD

**Issue by mtangoo** [thephpleague/oauth2-server#922]
I have been fighting with an error for quiet long time on System in production. The error it give is ` Client authentication failed`. Now The problem is, this is next to useless because many things can cause client to fail. I have tried to check all I could but I could not find what is wrong with client request so I decided to log the exception.

I have found that exception is not really useful as it contains no further details.

Can this library provide a way to get detailed exception message or at least provide way of logging stuffs so that troubleshooting does not become hellish? My life would have been easier if I could simply log detailed errors somehow. And I cannot put var_dump, since it is a live system

**Comment by simonhamp** [comment]
@mtangoo I agree that it's a bit tricky to spot mistakes in your client auth code...  during development. However, there are really only [3 cases](https://github.com/thephpleague/oauth2-server/search?q=invalidClient&type=Code) where the `invalid_client` exception is thrown, so unless you're doing something extra in your implementation, there's not many paths to investigate.

Also, the main consideration here is how you want something to behave in **production**: for extra safety, you **wouldn't** want to give away exactly what's failing to a bad actor as this will actually help them narrow their attack vectors.

I feel that the "right thing" to do here is **not** give the developer the ability to turn on some development-only insight and thereby open the door to accidentally leaving that on in production.

Generally, this issue will be caused by not passing the correct parameters or values in an auth request. If you can share the actual code that's building the request, I may be able to help further.

Have you tried building the request in a generic HTTP client such as Postman? This usually helps me to figure out where I'm going wrong because I can more easily see what's missing from the request. Further, eventually you will build up a collection of requests that look right and work well that you can reference in future.

**Comment by mtangoo** [comment]
>I feel that the "right thing" to do here is not give the developer the ability to turn on some development-only insight and thereby open the door to accidentally leaving that on in production.

Am not sure how other implements it, but my implementation which is Yii based, catches that exception, logs important stuffs and presents user with a generic message. Those detail can even be bundled as an exception hint if you still care for those who don't catch exceptions (which shouldn't be the case. It is developer responsibility to catch exceptions and do something)

Another way would be emitting an error Event with clear data of what happened. I see it is emiting the request, but that need one to dig the code. Emiting the actual error then the request will help a lot in logging, for those of us who care about it.

>Have you tried building the request in a generic HTTP client such as Postman? This usually helps me to figure out where I'm going wrong because I can more easily see what's missing from the request. 

What do you think of that?

As of the issue itself, I found the issue after (carefully) putting `var_dump` in the error and found client was sending wrong URL. Digging deeper, there was test config that was overriding the live parameters. So the problem itself is fixed, Thank you.

**Comment by simonhamp** [comment]
>Those detail can even be bundled as an exception hint if you still care for those who don't catch exceptions (which shouldn't be the case. It is developer responsibility to catch exceptions and do something)

Agreed! But there's also a balance of what's strictly **needed** to comply with [the spec here](https://tools.ietf.org/html/rfc6749#section-5.2) too.

Basically, the only required detail for the response is the actual `error` (in this case `invalid_client`) and that's what this package allows you to supply in your response when handling the exception.

If you choose to dig into this to provide the optional parameters in your response (e.g. `error_description`), I think that should be down to the developer and not a responsibility of the library.

I do believe that it would make things a little easier while developing if the cause of the error was surfaced, for sure. But ultimately it's not much gain for what I see two major drawbacks:

1. Potential for reduced security (as mentioned earlier)
2. Unnecessary extra processing required for cases where someone has typo'd some key detail - accuracy of all of those details should be the first thing to check on `invalid_client` errors

**Comment by mtangoo** [comment]
>Basically, the only required detail for the response is the actual error (in this case invalid_client) and that's what this package allows you to supply in your response when handling the exception.

Are you implying that somewhere in my implementation I can get the `actual error`?

**Comment by mtangoo** [comment]
> If you still feel that there is more needed from the library, please feel free to open a PR.

I still believe that adding actual error will benefit developers, with a little cost (that am not sure if matters anyway as they are only invoked in case of an error). But opening a PR at least should have a guarantee that it can be discussed. If adding that is beyond what is acceptable, then opening PR becomes waste :)
